homeassistant/components/zimi/__init__.py

```diff
@@ -0,0 +1,69 @@
+"""The zcc integration."""
+
+from __future__ import annotations
+
+import logging
+
+from zcc import ControlPoint, ControlPointError
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import CONF_HOST, CONF_PORT
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import ConfigEntryNotReady
+from homeassistant.helpers import device_registry as dr
+from homeassistant.helpers.device_registry import CONNECTION_NETWORK_MAC
+
+from .const import DOMAIN, PLATFORMS
+from .helpers import async_connect_to_controller
+
+_LOGGER = logging.getLogger(__name__)
+
+
+type ZimiConfigEntry = ConfigEntry[ControlPoint]
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ZimiConfigEntry) -> bool:
+    """Connect to Zimi Controller and register device."""
+
+    try:
+        api = await async_connect_to_controller(
+            host=entry.data[CONF_HOST],
+            port=entry.data[CONF_PORT],
+        )
+
+    except ControlPointError as error:
+        raise ConfigEntryNotReady(f"Zimi setup failed: {error}") from error
+
+    if not api:
+        raise ConfigEntryNotReady("Zimi setup failed: not ready")
+
+    _LOGGER.debug("\n%s", api.describe())
+
+    entry.runtime_data = api
+
+    device_registry = dr.async_get(hass)
+    device_registry.async_get_or_create(
+        config_entry_id=entry.entry_id,
+        identifiers={(DOMAIN, api.mac)},
+        manufacturer=api.brand,
+        name=f"{api.network_name}",
+        model="Zimi Cloud Connect",
+        sw_version=api.firmware_version,
+        connections={(CONNECTION_NETWORK_MAC, api.mac)},
+    )
+
+    await hass.config_entries.async_forward_entry_setups(entry, PLATFORMS)
+
+    _LOGGER.debug("Zimi setup complete")
+
+    return True
+
+
+async def async_unload_entry(hass: HomeAssistant, entry: ZimiConfigEntry) -> bool:
+    """Unload a config entry."""
+
+    api = entry.runtime_data
+    if api:
+        api.disconnect()
```

Comment:
This check doesn't seem needed. The `entry.runtime_data` is typed as `ControlPoint`.

```suggestion
    api.disconnect()
```

Reply:
I was trying to protect against the case when api wasn't set because the zcc api failed to connect and at the same time the user deleted the integration.    So I think there are cases where api is None and I need to check it's value first before accessing it's methods.

Reply:
No, `async_setup_entry` will raise `ConfigEntryNotReady` and the config entry state will become `ConfigEntryState.SETUP_RETRY`, if the zxx api failed to connect. `async_entry_unload` is only called if the config entry is loaded.